### ISSUE TITLE
Notes can't place to correct folder, sometimes.

### ISSUE THREAD

**Issue by kxsluna** [cweirup/JoplinSafariWebClipper#16]
Hi, thanks for you work!
I'm using Joplin 2.14.17 and latest clipper for safari, macOS 13.6.6.
Sometimes, the note just can't place to the right folder, but in some other folder. 
I have nested folders, there are some folders in its parent folder. My guess it is the reason.
So I try to find the pattern, like maybe always place to a folder above or below? But seems no pattern here.
Would you please fix this?

**Comment by cweirup** [comment]
So I fully understand: You are on a web page, you try to save it into Joplin, you have a nested folder selected, but after saving the page the clipping goes into a different folder. Is that accurate?

Other questions:
- Does the extension ever place the note in the correct folder? 
- Does it matter which way you are saving it? URL, Complete Page, Simplified Page?
- Does this only happen with nested folders or also with top-level folders? 
- Are all of your folders named differently or are there duplicate subfolder names, like under different parents?
- If you can share your Joplin folder structure, I can mimic that to see what happens on my end.

I can test out with additional nested folders. I did some testing with that, but admittedly missed some use cases. The extension needs to loop through arrays of the folders coming back from the API, so I wouldn't be surprised if there is some array traversing that needs to get fixed. 

Thanks!
Chris

**Comment by kxsluna** [comment]
Thanks for reply. Yes, that's the problem.

- Yes, The extension will work correctly sometime. Today, after a total reboot (system, softwares) it's fine now!
- The way saving the note seems not affect.
- Happen to all kind folders.
- My folder names contain Chinese characters, but there are no duplicate names, the structure like this:

> folder 1
> folder 2
> ...
> parent folder
>   -subfolder1
>   -subfolder2
> folder xx
> ...
> another parent folder
>   -subfolders
> folder yy
> ...

**Comment by cweirup** [comment]
I'll dig into this some more. I can't recreate this behavior on my set-up, but I will try it with Chinese characters. I don't know why that would cause an issue, however.

**Comment by kxsluna** [comment]
Finally! I found the way to recreate this problem:
1. Create a new notebook.
2. Drag some notebooks to the new book, so there is a nested notebook.
3. Then, clip webs will not place to the specific notebook, it will always show in the top folder of the notebooks list.
And I've tried to restart Joplin, not working. But restart Safari will do!

Hope these will help. Thanks!

**Comment by cweirup** [comment]
Ah, thanks, this does help. I need a bit of time to dig into this, but I suspect I'm caching the notebook tree and it doesn't refresh until you restart Safari.

Can you try one other thing? Do your steps 1 and 2. Then, rather than restarting Safari, can you disable the Joplin Clipper extension, then reenable the extension? I think that will force the extension to reload the notebook tree, which is essentially what you are doing when you restart Safari. Here is a screenshot of where to do that.

![Screenshot 2024-06-24 at 19 58 32](https://github.com/cweirup/JoplinSafariWebClipper/assets/216012/4c8ce52c-5fb5-40e7-8fd6-871f5ebe26d3)

If that works, then it will confirm I'm caching the notebook tree (or more accurately, due to the extension lifecycle, Safari is only re-rendering the window, not reloading the extension, where I load the notebook tree). I can look to see if I can check for Joplin status, but I suspect I'm not going to have a way to do that and would need to reload the notebook tree each time. I'll have to see how that impacts performance.

**Comment by kxsluna** [comment]
Yes, disable and enable the extension like you said will correct this problem.
Maybe, re-caching when exetsion menu shows up? 😄

**Comment by cweirup** [comment]
Thanks for the confirmation! Yes, I'll need to update how I refresh. Looks
like while I pull the notebook list each time, I'm not storing it correctly
each time. No timeline on when I can get to it, unfortunately, due to work.
